eudemo/eudemo/ivc/panelling/_paneller.py

```diff
@@ -0,0 +1,252 @@
+        """
+        Return the angles of rotation between each set of adjacent panels.
+
+        Note that this is the tail-tail angle between the panel vectors,
+        not the head-tail angles.
+        """
+        joints = self.joints(dists)
+        line_vectors = np.diff(joints, axis=1)
+        magnitudes = np.linalg.norm(line_vectors, axis=0)
+        # avoid division by zero later
+        magnitudes[np.isclose(magnitudes, 0)] += 1e-8
+        dots = (line_vectors[:, :-1] * line_vectors[:, 1:]).sum(axis=0)
+        dots /= magnitudes[:-1] * magnitudes[1:]
+        return np.degrees(np.arccos(np.clip(dots, -1.0, 1.0)), out=dots)
+
+    def panel_lengths(self, dists: np.ndarray) -> np.ndarray:
+        """Return the lengths of each panel."""
+        panel_vecs = np.diff(self.joints(dists))
+        return np.hypot(panel_vecs[0], panel_vecs[1])
+
+
+class LengthNormBoundary:
+    """Class to represent a wire interpolated over the normalised distance along it."""
+
+    def __init__(self, boundary_points: np.ndarray):
+        self.length_norm = norm_lengths(boundary_points)
+        self._x_spline = InterpolatedUnivariateSpline(
+            self.length_norm, boundary_points[0]
+        )
+        self._z_spline = InterpolatedUnivariateSpline(
+            self.length_norm, boundary_points[1]
+        )
+
+        self.tangent_norm = norm_tangents(boundary_points)
+        self._x_tangent_spline = InterpolatedUnivariateSpline(
+            self.length_norm, self.tangent_norm[0]
+        )
+        self._z_tangent_spline = InterpolatedUnivariateSpline(
+            self.length_norm, self.tangent_norm[1]
+        )
+
+    def x(self, dist: Union[float, np.ndarray]) -> np.ndarray:
+        """Find x at the given normalised distance along the boundary."""
+        return self._x_spline(dist)
+
+    def z(self, dist: Union[float, np.ndarray]) -> np.ndarray:
+        """Find z at the given normalised distance along the boundary."""
+        return self._z_spline(dist)
+
+    def x_tangent(self, dist: Union[float, np.ndarray]) -> np.ndarray:
+        """Find x at the tangent vector a given distance along the boundary."""
+        return self._x_tangent_spline(dist)
+
+    def z_tangent(self, dist: Union[float, np.ndarray]) -> np.ndarray:
+        """Find z at the tangent vector a given distance along the boundary."""
+        return self._z_tangent_spline(dist)
+
+
+def norm_lengths(points: np.ndarray) -> np.ndarray:
+    """
+    Calculate the cumulative normalized lengths between each 2D point.
+
+    Parameters
+    ----------
+    points
+        A numpy array of points, shape should be (2, N).
+    """
+    dists = np.diff(points, axis=1)
+    sq_dists = np.square(dists, out=dists)
+    summed_dists = np.sum(sq_dists, axis=0)
+    sqrt_dists = np.sqrt(summed_dists, out=summed_dists)
+    cumulative_sum = np.cumsum(summed_dists, out=sqrt_dists)
+    return np.hstack((0, cumulative_sum / cumulative_sum[-1]))
```

Comment:
What was the thoughts on using the `out` argument? 
It looks like we are performing operations and then renaming the variable without reassigning memory so after running L176 `summed_dists == sqrt_dists == cumlative_sum` . Unless I'm mistaken?

It seems a little confusing, but I assume you're trying to flag the logic flow?

Reply:
It's potentially premature optimisation, but it's for performance. Using `out` means we're not constantly re-allocating memory for new arrays; we're re-using memory we've already allocated.

I admit, when I first wrote this I thought it was going to be used within an optimisation loop, but it's not. Maybe I should remove the `out` arguments for clarity's sake, as I don't think performance matters much here?

Reply:
Possibly premature op but I don't mind. I have been bitten before using the same `out` array as the `in` array, you can get some weird bugs depending on the operations that are happening.

Reply:
This function has been removed, as there is a similar one in `geometry.coordinates`